docs/backend/api-definitions-guidelines.md

```diff
@@ -0,0 +1,20 @@
+# API Definitions Guidelines
+
+This document describes how service APIs will be managed for `InstructLab` and the sub-components of the `InstructLab` backend.
+
+## What parts of InstructLab need service APIs?
+
+There are two primary classes of service APIs needed to support InstructLab:
+
+* `Platform APIs`: These are APIs that are ignorant of `InstructLab` and provide generic AI platform capabilities (e.g., Fine Tuning, SDG, Eval)
+* `InstructLab APIs`: These are the APIs that reflect the user-facing functionality of `InstructLab` itself. They are aware of the end-to-end `InstructLab` workflow.
+
+The `InstructLab APIs` are essential for hosting `InstructLab` as a service in a repeatable way. The `Platform APIs` are critical for component reuse and extensibility (e.g., new SDG algorithms for new taxonomy data types), but they are not strictly required for hosting `InstructLab` as a service.
+
+## How will service APIs be defined?
+
+Service APIs will be defined using [OpenAPI](https://www.openapis.org/) format in [YAML](https://yaml.org/). For structural and style guidelines, see [api-definitions](../../api-definitions/README.md).
+
+## Where will service API definitions live?
+
+Service API definitions will live in [api-definitions](../../api-definitions) within this repo.
```

Comment:
If we put the API definition in here, we might have to rename this repository, it does not look like `dev-docs` anymore.
I'd personally be in favor of putting the definitions in a dedicated repo or as close to the code as possible (same repo as the code) for a few reasons:

* versioning, branching, and releasing a new API
* CI around validation and generation
* potential build and publishing of SDK (that's typically another repo too)

Reply:
Hmm, interesting @leseb. Should the APIs be defined in a specification repo? Its not REST but OpenTelemetry defines its [specifications](https://github.com/open-telemetry/opentelemetry-specification) in one place.

Reply:
@leseb Thanks for raising this. I too would prefer a dedicated repo, but thought it might be best to start here and move later. Depending on how quickly this ramps up, it may make more sense to just go straight to the end solution though?

The corresponding thought for me that is still unresolved is whether these API specs should all live under `instructlab` or not. The `platform` layer is going to intentionally NOT be `instructlab` specific so that these APIs can be used independently, but at the moment we don't have a "definition of the canonical AI service platform" name, so I don't know _where_ they should live. Having this question still open is part of why I figured a move was coming eventually and we might as well start small.

Definitely no strong opinions here for how we start, so I'm open to anything!

Reply:
W.r.t. the relationship between the API specs and the code, I agree that there are really three possible options:

1. The API lives with the implementation
  * Pro: This is the easiest to navigate from a project-centric view
  * Con: It's much harder to keep commonality across different services/projects and it's much harder to treat the API as an abstract definition if we decide to reimplement it (or maintain multiple implementations for different deployment scenarios)
3. The API lives in a central place and is used by-reference in the implementation
  * Pro: Single source of truth, easy to see platform-centric view
  * Con: This is the hardest to navigate from a project perspective (submodules, fetch scripts, yuck!) 
5. The API lives is a central place as the "source of truth," and a copy (possibly with additional features) lives by-value with the implementation
  * Pro: This gives the best of both from a user perspective (easy to see project view, easy to see platform view)
  * Con: There is no longer a single source-of-truth and the two copies must be kept in sync

Personally, I've tended towards (2) in the past, but the logistical headaches of the `submodule` approach have been a nightmare. We also attempted to solve (2) with the source-of-truth living [in code](https://github.com/caikit/caikit/tree/main/caikit/interfaces), but the `code -> API spec` auto-generation turned out to be just as big a nightmare and folks continuously complained that it was impossible to actually _see_ the API.

For "move fast and break things," (1) is clearly the best approach, but I think the phase we're at right now is to try to back off from this perspective and create some commonality that is divorced from the individual project implementations.

I'm currently leaning towards (3) for these reasons. This would mean that the individual projects _could_ (should?) keep their own copies of their respective APIs, but that the "platform definition" would live in the central location and the onus would be on the project owner to keep the project in sync with the platform (i.e. contribute changes to the platform to support changes at the project level).

Reply:
The other option is we can start with 1, since it will be changing a lot initially, once we stabilize we can look into option 3.

Reply:
> (2) with the source-of-truth living [in code](https://github.com/caikit/caikit/tree/main/caikit/interfaces), but the code -> API spec auto-generation turned out to be just as big a nightmare and folks continuously complained that it was impossible to actually see the API.

You could generate api definitions and store them along side the implementation, for user consumption. In one of my pet projects, I [use](https://github.com/booxter/letsrolld/blob/a7c0fca2196eff33308f565520da5c3f00b5ddd0/src/letsrolld/webapi/app.py#L233) [flask-restful-swagger-3](https://pypi.org/project/flask-restful-swagger-3/) that is used to enforce compliance in runtime as well as to generate swagger definitions (that I then store in the same repo - the correspondence between the two is enforced by github checks.)

Reply:
The choice also depends on the programming language's dependency management capabilities. For example, with Go, option 2 is straightforward and avoids the hassle of submodules. You can have one repository as the source of truth for the API definition. Then, a separate repository hosts the generated client API code. The main Go project can use this client library as a module dependency. A similar approach could work with Python, where you produce a package on PyPI, thus avoiding submodules.

Option 3 with Git is essentially a re-sync, which is fairly easy to manage. However, this approach mimics what a submodule would do, requiring extra effort for synchronization. If we can devise an efficient synchronization strategy, this option might be viable.

I still prefer option 2 despite its management overhead because it provides a clear separation of concerns and an obvious location for the source of truth.

Option 1 is impractical since we will likely have multiple components consuming the API.

Reply:
@leseb Interesting point about generating packages off of the central repo. One of the things I've always liked about both OpenAPI and Protobuf is that the specification does not mandate an implementation language, so the server is free to implement the spec in any language they choose. I think we could accomplish this with your suggestion of a central repo if that repo's CI produced and published packages for all of the "well supported" languages (probably `python` + `go` to start, possibly with `rust` per [the other thread in Russell's proposal](https://github.com/instructlab/dev-docs/pull/74#discussion_r1630126956).

If we were to go this route, I think it would be a strong push to NOT hold these API definitions in the `dev-docs` repo so that we weren't mixing CI concerns between building docs and building generated code packages.

Reply:
One repository as source of truth for the specifications would be my preference. The `dev-docs` is not a suitable locations as said.

Reply:
Resolving this thread since we 're all on the same page!